### ISSUE TITLE
Generic server terminating due to "key exists"

### ISSUE THREAD

**Issue by lukas-krecan-lt** [rabbitmq/rabbitmq-server#1153]
We are getting lot of errors like this. 

=ERROR EVENT====
** Reason for termination == 
** {{key_exists,4},

Rabbit 3.6.6 Erlang 19.1

Can you please advice as what might be the reason?

```
=ERROR EVENT==== Tue, 21 Mar 2017 13:18:52 GMT ===
** Generic server <0.14800.1989> terminating
** Last message in was {{'basic.deliver',
                            <<"amq.ctag-PYrFsDu432cCG7sqjZOoLQ">>,11,false,
                            <<"amq.topic">>,<<"300339828_P">>},
                        {amqp_msg,
                            {'P_basic',undefined,undefined,undefined,
                                undefined,undefined,undefined,undefined,
                                <<"180000">>,undefined,undefined,undefined,
                                undefined,undefined,undefined},
                            <<"adler.taxi.data.push.MyPositionPush;\n{\"orderId\":\"902495032\",\"timestamp\":\"2017-03-21T13:16:21.654Z\",\"myPosition\":{\"_type\":\"adler.taxi.data.push.v1.SimplePosition\",\"lang\":null,\"placeName\":null,\"lat\":50.0638189,\"coordinates\":null,\"lon\":14.5160094,\"timestamp\":\"2017-03-21T13:16:21.000Z\",\"precision\":6},\"userProfileId\":\"800003812\"}">>},
                        {<0.4821.2084>,<0.751.1974>,<0.18028.1979>}}
** When Server state == {state,#Port<0.45074739>,
                         "46.135.29.232:55172 -> 10.58.54.3:1883",true,
                         undefined,true,running,
                         {none,<0.22140.2047>},
                         <0.3295.2000>,false,none,
                         {proc_state,#Port<0.45074739>,
                          {dict,1,16,16,8,80,48,
                           {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                           {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],
                             [["300339828_P",1]],
                             []}}},
                          {undefined,<<"amq.ctag-PYrFsDu432cCG7sqjZOoLQ">>},
                          {0,nil},
                          {9,
                           {4,4,
                            {2,2,nil,{3,3,nil,nil}},
                            {6,6,
                             {5,5,nil,nil},
                             {7,7,nil,{8,8,nil,{9,9,nil,{10,10,nil,nil}}}}}}},
                          2,4,"300339828_P",false,
                          {mqtt_msg,false,1,"lwt-presence",false,undefined,
                           <<"D 300339828_P">>},
                          {<0.4821.2084>,<0.19525.2105>},
                          <0.20533.1938>,<<"amq.topic">>,
                          {amqp_adapter_info,
                           {10,58,54,3},
                           1883,
                           {46,135,29,232},
                           55172,<<"46.135.29.232:55172 -> 10.58.54.3:1883">>,
                           {'MQTT',"N/A"},
                           [{channels,1},
                            {channel_max,1},
                            {frame_max,0},
                            {client_properties,
                             [{<<"product">>,longstr,<<"MQTT client">>}]},
                            {ssl,false}]},
                          none,<0.2133.0>,
                          {auth_state,<<"mqtt-client">>,
                           {user,<<"mqtt-client">>,[],
                            [{rabbit_auth_backend_internal,none}]},
                           <<"mqtt">>},
                          #Fun<rabbit_mqtt_processor.0.126118796>},
                         <0.20533.1938>,
                         {state,fine,5000,#Ref<0.0.3889168385.213327>}}
** Reason for termination == 
** {{key_exists,4},
    [{gb_trees,insert_1,4,[{file,"gb_trees.erl"},{line,318}]},
     {gb_trees,insert,3,[{file,"gb_trees.erl"},{line,277}]},
     {rabbit_mqtt_processor,amqp_callback,2,
                            [{file,"src/rabbit_mqtt_processor.erl"},
                             {line,357}]},
     {rabbit_mqtt_reader,handle_info,2,
                         [{file,"src/rabbit_mqtt_reader.erl"},{line,103}]},
     {gen_server2,handle_msg,2,[{file,"src/gen_server2.erl"},{line,1032}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
```

**Comment by michaelklishin** [comment]
Thank you for your time.

Team RabbitMQ uses GitHub issues for specific actionable items engineers can work on. This assumes we have a certain amount of information to work with. Questions, investigations, root cause analysis, discussions for potential features are all considered to be mailing list material by our team. When/if we have enough details and evidence we'd be happy to file a new issue.

Please post this to [rabbitmq-users](groups.google.com/forum/#!forum/rabbitmq-users). Thank you.

**Comment by michaelklishin** [comment]
The exception comes from the MQTT plugin which seems to run into duplicate message IDs. I am not aware of any known scenarios when that might happen. Please post a way to reproduce to rabbitmq-users and if it turns out to be an actionable issue in the plugin, we will file a new issue on your behalf. Thanks.

**Comment by lukas-krecan-lt** [comment]
Thanks. We were sending messages using AMQP and receiving them using MQTT and did not use message ID. Adding message ID to AMQP publish helped (so far).

**Comment by michaelklishin** [comment]
Interesting, we should at least add it to the docs. Thank you for reporting back, we will investigate and file new issues as needed (or just update the docs directly).

**Comment by lukas-krecan-lt** [comment]
I am still not sure that it really helped, I will confirm it next week.

**Comment by lukas-krecan-lt** [comment]
Sorry, setting message ID did not help

**Comment by saterHATER** [comment]
If I'm not mistaken, this is a working link to [rabbitMQ-users](https://groups.google.com/forum/?nomobile=true#!searchin/rabbitmq-users/src$2Frabbit_mqtt_processor.erl%7Csort:date/rabbitmq-users/qDpLQwDmA-A/QlzUbZymAAAJ). My apologies Michael I don't think your link was working.

**Comment by lukebakken** [comment]
I fixed this issue in rabbitmq/rabbitmq-mqtt#149. This fix will ship in the next RabbitMQ minor release (`3.6.15`)

**Comment by michaelklishin** [comment]
Duplicate of https://github.com/rabbitmq/rabbitmq-mqtt/issues/132.
